cmd/charmd/main.go

```diff
@@ -95,40 +94,17 @@ func serve(confPath string) error {
 			MaxAge:   conf.AuditLogMaxAge,
 		}
 	}
-
-	ring := bakery.NewPublicKeyRing()
-	if conf.IdentityPublicKey != nil {
-		err = ring.AddPublicKeyForLocation(cfg.IdentityLocation, false, conf.IdentityPublicKey)
-		if err != nil {
-			return errgo.Mask(err)
-		}
-	} else {
-		pubKey, err := httpbakery.PublicKeyForLocation(http.DefaultClient, cfg.IdentityLocation)
-		if err != nil {
-			return errgo.Mask(err)
-		}
-		err = ring.AddPublicKeyForLocation(cfg.IdentityLocation, false, pubKey)
-		if err != nil {
-			return errgo.Mask(err)
-		}
+	cache := bakery.NewPublicKeyRing()
+	err = addPublicKey(cache, conf.IdentityLocation, conf.IdentityPublicKey)
+	if err != nil {
+		return errgo.Mask(err)
 	}
-	if conf.TermsPublicKey != nil {
-		err = ring.AddPublicKeyForLocation(cfg.TermsLocation, false, conf.TermsPublicKey)
-		if err != nil {
-			return errgo.Mask(err)
-		}
-	} else if cfg.TermsLocation != "" {
-		pubKey, err := httpbakery.PublicKeyForLocation(http.DefaultClient, cfg.TermsLocation)
-		if err != nil {
-			return errgo.Mask(err)
-		}
-		err = ring.AddPublicKeyForLocation(cfg.TermsLocation, false, pubKey)
-		if err != nil {
-			return errgo.Mask(err)
-		}
+	err = addPublicKey(cache, conf.TermsLocation, conf.TermsPublicKey)
+	if err != nil {
+		return errgo.Mask(err)
 	}
 
-	cfg.PublicKeyLocator = ring
+	cfg.PublicKeyLocator = httpbakery.NewPublicKeyRing(http.DefaultClient, cache)
```

Comment:
If you're going to call httpbakery.PublicKeyForLocation in addPublicKey, then there's no need to use httpbakery.NewPublicKeyRing. The main thing that httpbakery.NewPublicKeyRing does is to fetch public keys when needed, which you don't need if you're going to pre-populate the ring.